### ISSUE TITLE
Support checking all files' text encodings, if declared to be UTF-8

### ISSUE THREAD

**Issue by sanmai-NL** [editorconfig-checker/editorconfig-checker#40]
Enforcing a specified text encoding in a monorepo rife with all kinds of textual resources is hard for crossplatform development teams. This feature would be of great value, and simple to implement. Especially if only a BOM-check is done. Alternatively, every text file that is covered by the EditorConfig settings can be decoded as UTF-8 on the fly.

**Comment by mstruebing** [comment]
Hey, I will look into it soonish.
Can you say something about BOM-checks? I've never head of it.

**Comment by sanmai-NL** [comment]
Great!

The BOM is explained at https://en.wikipedia.org/wiki/Byte_order_mark. See `charset` parameter value `utf-8-bom` at https://github.com/editorconfig/editorconfig/wiki/EditorConfig-Properties. Using this is more efficient than decoding entire files.

**Comment by mstruebing** [comment]
I've just played around with it a little and while coding and researching I've found this: `Use of a BOM is neither required nor recommended for UTF-8`: https://stackoverflow.com/questions/2223882/whats-the-difference-between-utf-8-and-utf-8-without-bom.

I've tested all files of this repository and not one has a BOM.
So I don't want to implement not recommended things and I can not reproduce the use case.

Don't get me wrong, I would want to have charset detection, but I would want to do it in a proper and reliable way.

Are there any other ways to get a files charset?

**Comment by sanmai-NL** [comment]
@mstruebing:
I see why you got that impression.

A lot of people, esp. Unix-fans, are very vocal about UTF-8 and BOMs. I see the UTF-8 BOM as a pragmatic solution to a serious problem. Some people cling to older software (that cannot decode UTF-8) and niche workflows (specific fragile combinations of text-based Unix CLI tools). Others, like me are carefully adapting to the conventions applied by popular operating systems and software such as Microsoft Windows and Office. Also, [don't be fooled by the wording in that quote](https://stackoverflow.com/questions/2223882/whats-the-difference-between-utf-8-and-utf-8-without-bom/2223926#comment57102871_2223926). It doesn't mean UTF-8 BOM is bad at all. It's very useful.

You can also check whether the text encoding is UTF-8 by decoding the file in full (reading it as UTF-8 string).

**Comment by rasa** [comment]
Fixed in #284. Please close this ticket.

**Comment by cfbao** [comment]
@rasa tested a build from the main branch. Doesn't seem to be fixed.

A folder with `charset = utf-8` specified in .editorconfig, and a file encoded with utf-8-bom, editorconfig-checker didn't catch this violation.

**Comment by rasa** [comment]
@cfbao I'm sorry to hear that. I wrote #204, and the tests don't [check](https://github.com/editorconfig-checker/editorconfig-checker/blob/6a81239415a6d3a4e41a753d2f30b5db73db32d9/pkg/encoding/encoding_test.go#L41) for the BOM marker. We can add that.

Our tests do [check](https://github.com/editorconfig-checker/editorconfig-checker/blob/6a81239415a6d3a4e41a753d2f30b5db73db32d9/pkg/encoding/encoding_test.go#L62) for UTF16, so I'm surprised the code doesn't. I will look into that as well.

Thanks for the follow up.

**Comment by sanmai-NL** [comment]
https://en.wikipedia.org/wiki/Byte_order_mark?wprov=sfti1

**Comment by rasa** [comment]
@cfbao As you can see, I tried to fix this issue, but the chardet package we use misidentifies charsets, so we need to fix that package, or switch to a different package.

**Comment by sanmai-NL** [comment]
Detecting character sets may be too fine-grained and unreliable at the same time. Simply decoding the bytes as ASCII, when that fails as UTF-8 may be enough.

**Comment by sanmai-NL** [comment]
These encodings are supported by the standard:
> charset: set to latin1, utf-8, utf-8-bom, utf-16be or utf-16le to control the character set.

**Comment by rasa** [comment]
@mstruebing @klaernie What about using cgo and linking to https://gitlab.freedesktop.org/uchardet/uchardet ? The zig compiler may help make this relatively painless.

**Comment by klaernie** [comment]
To me this sounds like complicating the build process enormously just to use a library written in C. I'd personally rather choose a library written in go, since that also reduces the friction for compiling on something else besides Linux.

**Comment by rasa** [comment]
> To me this sounds like complicating the build process enormously just to use a library written in C. I'd personally rather choose a library written in go, since that also reduces the friction for compiling on something else besides Linux.

I agree, but I haven't found a go package  that reliably determines a file's encoding. If you know of one, please suggest it.

**Comment by rasa** [comment]
https://github.com/wlynxg/chardet looks very promising. If it works as I expect, I will submit this PR for review.

**Comment by klaernie** [comment]
> https://github.com/wlynxg/chardet looks very promising. If it works as I expect, I will submit this PR for review.

That really looks good, and I like to that it is being maintained. Great find!
